inst/include/dust/dust.hpp

```diff
@@ -141,8 +181,66 @@ class filter_state {
   std::vector<real_t> history_value;
   std::vector<size_t> history_order;
 };
+
+template <typename real_t>
+struct device_state {
+  void initialise(size_t n_particles, size_t n_state) {
+    const size_t n_rng = dust::rng_state_t<real_t>::size();
+    // NOTE: not setting up yi_selected here, which was used in dustgpu
+    y = dust::device_array<real_t>(n_state * n_particles);
+    rng = dust::device_array<uint64_t>(n_rng * n_particles);
+  }
+  dust::device_array<real_t> y;
+  dust::device_array<uint64_t> rng;
+};
+
+// We need to compute the size of space required for integers and
+// reals on the device, per particle. Because we work on the
+// requirement that every particle has the same dimension we pass an
+// arbitrary set of shared parameters (really the first) to
+// device_work_size. The underlying model can overload this template
+// for either real or int types and return the length of data
+// required.
+template <typename T>
+size_t device_work_size_int(typename dust::shared_ptr<T> shared) {
+  return 0;
+}
+
+template <typename T>
+size_t device_work_size_real(typename dust::shared_ptr<T> shared) {
+  return 0;
+}
+
 }
 
+// We'll need to expand this soon to cope with shared memory, but that
+// will be coming via another change to dust. Or we can do it here
+// with a shared object that contains just:
+// template <typename T>
+// struct shared_t {
+//   const real_t const * real_data;
+//   const int * const * int_data;
+//   // plus lengths?
+// };
+//
+// which can just point at the data in the vector because it will live
+// longer than the object.
+template <typename T>
+void update_device(size_t step,
+                   const dust::interleaved<typename T::real_t> state,
+                   dust::interleaved<int> internal_int,
+                   dust::interleaved<typename T::real_t> internal_real,
+                   dust::shared_ptr<T> shared,
+                   dust::rng_state_t<typename T::real_t>& rng_state,
+                   dust::interleaved<typename T::real_t> state_next);
+
+template <typename T>
+void run_particles(size_t step_from, size_t step_to, size_t n_particles,
+                   size_t n_state, size_t n_int, size_t n_real, size_t n_pars,
+                   typename T::real_t * state,
+                   std::vector<dust::shared_ptr<T>> shared,
```

Comment:
This will need to be something simple (number/pointer) as this is the global/kernel signature. I suspect we may have:
```
real_t * shared_real
int * shared_int
int * shared_index
```
Shared_index is probably just the pars index vector already in the dust object